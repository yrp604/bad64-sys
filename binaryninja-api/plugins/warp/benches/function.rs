use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use criterion::{criterion_group, criterion_main, Criterion};
use rayon::prelude::*;
use std::path::PathBuf;
use warp_ninja::build_function;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 1] = ["atox.obj"];

pub fn function_benchmark(c: &mut Criterion) {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");

        let functions: Vec<_> = view.functions().iter().map(|f| f.to_owned()).collect();
        // Bench all functions sequentially
        c.bench_function("signature all functions", |b| {
            b.iter(|| {
                for func in &functions {
                    let _ = build_function(&func, || func.lifted_il().ok(), false);
                }
            })
        });

        // Bench all functions in parallel
        c.bench_function("signature all functions rayon", |b| {
            b.iter(|| {
                functions
                    .par_iter()
                    .filter_map(|func| {
                        Some(build_function(
                            func.as_ref(),
                            || func.lifted_il().ok(),
                            false,
                        ))
                    })
                    .collect::<Vec<_>>()
            })
        });
    }
}

criterion_group!(benches, function_benchmark);
criterion_main!(benches);
