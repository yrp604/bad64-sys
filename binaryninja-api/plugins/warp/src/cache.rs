pub mod container;
pub mod function;
pub mod guid;
pub mod type_reference;

pub use function::*;
pub use guid::*;
pub use type_reference::*;

use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::function::Function as BNFunction;
use binaryninja::rc::Guard;
use binaryninja::rc::Ref as BNRef;
use binaryninja::ObjectDestructor;
use std::hash::{DefaultHasher, Hash, Hasher};

pub fn register_cache_destructor() {
    pub static mut CACHE_DESTRUCTOR: CacheDestructor = CacheDestructor;
    #[allow(static_mut_refs)]
    // SAFETY: This can be done as the backing data is an opaque ZST.
    unsafe {
        CACHE_DESTRUCTOR.register()
    };
}

/// A unique view ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct ViewID(u64);

impl From<&BinaryView> for ViewID {
    fn from(value: &BinaryView) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.original_image_base());
        hasher.write_usize(value.file().session_id());
        Self(hasher.finish())
    }
}

impl From<BNRef<BinaryView>> for ViewID {
    fn from(value: BNRef<BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BinaryView>> for ViewID {
    fn from(value: Guard<'_, BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

/// A unique function ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct FunctionID(u64);

impl From<&BNFunction> for FunctionID {
    fn from(value: &BNFunction) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.start());
        hasher.write_u64(value.lowest_address());
        hasher.write_u64(value.highest_address());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNFunction>> for FunctionID {
    fn from(value: BNRef<BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNFunction>> for FunctionID {
    fn from(value: Guard<'_, BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

pub struct CacheDestructor;

impl ObjectDestructor for CacheDestructor {
    fn destruct_view(&self, view: &BinaryView) {
        clear_type_ref_cache(view);
        log::debug!("Removed WARP caches for {:?}", view.file().filename());
    }
}
