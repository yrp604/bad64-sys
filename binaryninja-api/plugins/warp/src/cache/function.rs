use crate::convert::{comment_to_bn_comment, to_bn_symbol_at_address};
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::function::{Function as BNFunction, FunctionUpdateType};
use binaryninja::symbol::SymbolType;
use warp::signature::function::Function;

// TODO: Rename this?
/// Inserts a function match into the cache. This also has the side effect of setting persisted function
/// information, such as the matched function symbol.
///
/// IMPORTANT: This will mark the function as needing updates, if you intend to fill in functions with
/// no match (i.e. `None`), then you must change this function to prevent marking that as needing updates.
/// However, it's perfectly valid to remove a match and need to update the function still, so be careful.
pub fn insert_cached_function_match(function: &BNFunction, matched_function: Option<Function>) {
    let view = function.view();
    let function_start = function.start();
    // NOTE: If we expect to run match_function multiple times on a function, we should move this elsewhere.
    // Mark the function as needing updates so that reanalysis occurs on the function, and we apply the match.
    function.mark_updates_required(FunctionUpdateType::FullAutoFunctionUpdate);
    if let Some(auto_sym) = view.symbol_by_address(function_start) {
        // TODO: If we ever create non library function symbols we will need to remove this check (see: `to_bn_symbol_at_address`).
        if auto_sym.sym_type() == SymbolType::LibraryFunction {
            // NOTE: This will also mark for full auto function update, one thing to note is that the
            // requirement to call this is that this function not be called in the associated function's analysis.
            view.undefine_auto_symbol(&auto_sym);
        }
    }
    match matched_function {
        Some(matched_function) => {
            // Define the new matched function symbol, this can safely be done here as the symbol itself
            // will be persisted, unlike function type information or variable information.
            let new_sym = to_bn_symbol_at_address(&view, &matched_function.symbol, function_start);
            view.define_auto_symbol(&new_sym);
            // TODO: How to clear the comments? They are just persisted.
            // TODO: Also they generate an undo action, i hate implicit undo actions so much.
            for comment in &matched_function.comments {
                let bn_comment = comment_to_bn_comment(&function, comment.clone());
                function.set_comment_at(bn_comment.addr, &bn_comment.comment);
            }
            function.store_metadata("warp_matched_function", &matched_function.to_bytes(), false);
        }
        None => {
            function.remove_metadata("warp_matched_function");
        }
    }
}

// TODO: This does allocations, and for every reanalysis.
pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let metadata = function.query_metadata("warp_matched_function")?;
    let raw_metadata = metadata.get_raw()?;
    Function::from_bytes(&raw_metadata)
}
