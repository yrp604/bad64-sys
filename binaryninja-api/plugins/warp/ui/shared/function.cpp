#include "theme.h"

#include "function.h"

#include <QClipboard>
#include <QGridLayout>
#include <QHeaderView>

#include "constraint.h"
#include "misc.h"

WarpFunctionItem::WarpFunctionItem(Warp::Ref<Warp::Function> function,
                                   BinaryNinja::Ref<BinaryNinja::Function> analysisFunction)
{
    m_function = function;

    BinaryNinja::Ref<BinaryNinja::Symbol> symbol = m_function->GetSymbol(*analysisFunction);
    std::string symbolName = symbol->GetShortName();
    setText(QString::fromStdString(symbolName));

    // Serialize the tokens to make it accessible via QModelIndex.
    // We will take these tokens and then user them in our custom item delegate.
    TokenData tokenData = TokenData(symbolName);
    if (BinaryNinja::Ref<BinaryNinja::Type> type = m_function->GetType(*analysisFunction))
        tokenData = TokenData(*type, symbolName);
    setData(QVariant::fromValue(tokenData), Qt::UserRole);
}

void WarpFunctionItem::SetContainer(const Warp::Ref<Warp::Container> &container)
{
    m_container = container;

    // Add the container string to data so the filter model picks it up.
    auto containerName = m_container->GetName();
    setData(QString::fromStdString(containerName), Qt::UserRole + 2);
}

void WarpFunctionItem::SetSource(Warp::Source source)
{
    m_source = source;

    // Add the source string to data so the filter model picks it up.
    std::string sourceStr = m_source->ToString();
    setData(QString::fromStdString(sourceStr), Qt::UserRole + 1);
}

WarpFunctionItemModel::WarpFunctionItemModel(const QStringList &labels, QObject *parent)
{
    this->setHorizontalHeaderLabels(labels);
}

void WarpFunctionItemModel::AppendFunction(WarpFunctionItem *item)
{
    QList<QStandardItem *> row = {};
    row.insert(COL_FUNCTION_ITEM, item);
    appendRow(row);
}

void WarpFunctionItemModel::InsertFunction(uint64_t address, WarpFunctionItem *item)
{
    // Update item if already available, this lets us keep the model
    const auto iter = m_insertableFunctionRows.find(address);
    if (iter != m_insertableFunctionRows.end())
    {
        setItem(iter->second, COL_FUNCTION_ITEM, item);
        return;
    }

    AppendFunction(item);
    m_insertableFunctionRows[address] = rowCount() - 1;
}

WarpFunctionItem *WarpFunctionItemModel::GetItem(const QModelIndex &index) const
{
    if (!index.isValid())
        return nullptr;
    return dynamic_cast<WarpFunctionItem *>(item(index.row(), COL_FUNCTION_ITEM));
}

std::optional<uint64_t> WarpFunctionItemModel::GetAddress(const QModelIndex &index) const
{
    if (!index.isValid())
        return std::nullopt;
    // TODO: This is a hack, this means we must enumerate all rows to get the address.
    for (const auto &[addr, row]: m_insertableFunctionRows)
        if (row == index.row())
            return addr;
    return std::nullopt;
}

QVariant WarpFunctionItemModel::data(const QModelIndex &index, int role) const
{
    if (role == Qt::BackgroundRole)
    {
        auto itemFunction = GetItem(index);
        // Check if we have a valid item and it's the matched function
        if (m_matchedFunction && itemFunction)
        {
            // TODO: Why wont == go to the correct call???
            if (BNWARPFunctionsEqual(itemFunction->GetFunction()->m_object, m_matchedFunction->m_object))
            {
                // TODO: Better color?
                static QColor matchedColor = getThemeColor(BlueStandardHighlightColor);
                matchedColor.setAlpha(128);
                return matchedColor;
            }
        }
    }

    if (role == Qt::DisplayRole)
    {
        // We really only use this for searching as we have TokenData for our delegate.
        WarpFunctionItem *item = GetItem(index);
        if (!item)
            return QVariant();
        TokenData tokenData = item->data(Qt::UserRole).value<TokenData>();
        // Add the function guid so we can filter by that.
        QString text = tokenData.toString() + " " + QString::fromStdString(item->GetFunction()->GetGUID().ToString());
        if (auto source = item->GetSource(); source)
        {
            // Add the source guid so we can also filter by that.
            std::string sourceStr = source->ToString();
            text = text + " " + QString::fromStdString(sourceStr);
        }
        return text;
    }

    return QStandardItemModel::data(index, role);
}

bool WarpFunctionFilterModel::filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const
{
    const QString filterString = filterRegularExpression().pattern();
    if (filterString.isEmpty())
        return true;

    // Filter on the first column only, this contains our actual function.
    auto index = sourceModel()->index(sourceRow, 0, sourceParent);
    auto data = QRegularExpression::escape(index.data().toString());
    if (data.contains(filterString, Qt::CaseInsensitive))
        return true;
    return false;
}

bool WarpFunctionFilterModel::lessThan(const QModelIndex &sourceLeft, const QModelIndex &sourceRight) const
{
    // TODO: When we make the stuff _actually_ sortable.
    return sourceLeft.row() < sourceRight.row();
}

WarpFunctionTableWidget::WarpFunctionTableWidget(QWidget *parent) : QWidget(parent)
{
    QGridLayout *layout = new QGridLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(2);

    m_table = new QTableView(this);
    m_model = new WarpFunctionItemModel({"Function"}, this);
    m_proxyModel = new WarpFunctionFilterModel(this);
    m_proxyModel->setSourceModel(m_model);
    m_table->setModel(m_proxyModel);

    m_filterEdit = new FilterEdit(this);
    m_filterView = new FilteredView(this, m_table, this, m_filterEdit);
    m_filterView->setFilterPlaceholderText("Search functions (By GUID, name or source)");

    layout->addWidget(m_filterEdit, 0, 0, 1, 5);
    layout->addWidget(m_table, 1, 0, 1, 5);

    // Make the table look nice.
    m_table->horizontalHeader()->setStretchLastSection(true);
    m_table->verticalHeader()->hide();
    m_table->setSelectionBehavior(QAbstractItemView::SelectRows);
    m_table->setSelectionMode(QAbstractItemView::SingleSelection);
    m_table->setEditTriggers(QAbstractItemView::NoEditTriggers);
    m_table->setFocusPolicy(Qt::NoFocus);
    m_table->setShowGrid(false);
    m_table->setAlternatingRowColors(false);
    m_table->setSortingEnabled(true);
    // NOTE: We only have a single column right now, so disable header.
    m_table->horizontalHeader()->hide();
    // Decrease row height to make it look nice.
    m_table->verticalHeader()->setDefaultSectionSize(30);

    // Make the highlight less bright.
    QPalette palette = m_table->palette();
    palette.setColor(QPalette::Highlight, getThemeColor(SelectionColor));
    m_table->setPalette(palette);

    TokenDataDelegate *tokenDelegate = new TokenDataDelegate(this);
    // NOTE: Column 0 is assumed to be the function with the token data.
    m_table->setItemDelegateForColumn(0, tokenDelegate);

    AddressColorDelegate *addressDelegate = new AddressColorDelegate(this);
    // NOTE: Column 1 is assumed to be the function address.
    m_table->setItemDelegateForColumn(1, addressDelegate);

    // Add a dynamic context menu to the table.
    // NOTE: This is a bit stupid, I am sure there is a better way to do this in QT.
    m_contextMenu = new QMenu(this);
    RegisterContextMenuAction("Copy Name", [](WarpFunctionItem *item, std::optional<uint64_t>) {
        QClipboard *clipboard = QGuiApplication::clipboard();
        clipboard->setText(item->text());
    });
    RegisterContextMenuAction("Copy GUID", [](WarpFunctionItem *item, std::optional<uint64_t>) {
        QClipboard *clipboard = QGuiApplication::clipboard();
        Warp::Ref<Warp::Function> function = item->GetFunction();
        std::string guidStr = function->GetGUID().ToString();
        clipboard->setText(QString::fromStdString(guidStr));
    });

    m_table->setContextMenuPolicy(Qt::CustomContextMenu);
    connect(m_table, &QTableView::customContextMenuRequested, this, [&](QPoint pos) {
        const QModelIndex index = m_table->indexAt(pos);
        if (!index.isValid())
            return;
        const QModelIndex sourceIndex = m_proxyModel->mapToSource(index);
        WarpFunctionItem *item = m_model->GetItem(sourceIndex);
        if (!item || !item->GetFunction())
            return;

        // Execute the menu and get the selected action
        const QAction *selectedAction = m_contextMenu->exec(m_table->viewport()->mapToGlobal(pos));
        if (!selectedAction)
            return;

        const auto name = selectedAction->text();
        const auto iter = m_contextMenuActions.find(name);
        if (iter != m_contextMenuActions.end())
            iter->second(item, m_model->GetAddress(sourceIndex));
    });
}

void WarpFunctionTableWidget::RegisterContextMenuAction(const QString &name,
                                                        const std::function<void(
                                                            WarpFunctionItem *, std::optional<uint64_t>)> &callback)
{
    m_contextMenu->addAction(name);
    m_contextMenuActions[name] = callback;
}

void WarpFunctionTableWidget::SetFunctions(QVector<WarpFunctionItem *> functions)
{
    // Clear matches as they are no longer valid.
    m_model->clear();
    m_model->setRowCount(0);

    // Temporarily disable sorting so we can add rows faster
    m_table->setModel(m_model);
    m_table->setSortingEnabled(false);
    m_table->setEnabled(false);

    for (const auto &function: functions)
        m_model->AppendFunction(function);

    // We are done, re-enable table.
    m_table->setEnabled(true);
    m_table->setModel(m_proxyModel);
    m_table->setSortingEnabled(true);

    // Update the filter text with the new count of functions.
    m_filterView->setFilterPlaceholderText(QString("Search %1 functions").arg(m_model->rowCount()));
}

void WarpFunctionTableWidget::InsertFunction(uint64_t address, WarpFunctionItem *function)
{
    m_model->InsertFunction(address, function);
}

void WarpFunctionTableWidget::setFilter(const std::string &filter)
{
    m_proxyModel->setFilterFixedString(QString::fromStdString(filter));
    m_filterView->showFilter(QString::fromStdString(filter));
}

WarpFunctionInfoWidget::WarpFunctionInfoWidget(QWidget *parent)
    : QWidget(parent)
{
    // Create a tab widget
    QTabWidget *tabWidget = new QTabWidget(this);
    tabWidget->setContentsMargins(0, 0, 0, 0);

    // Create tables for the "Constraints", "Comments", and "Variables" tabs
    m_commentsTable = new QTableView(this);
    // m_variablesTable = new QTableView(this);

    // TODO: On click navigate to where the constraint is located.
    m_constraintsTable = new WarpConstraintTableWidget(this);
    tabWidget->addTab(m_constraintsTable, "Constraints");

    // Set up comments tab
    m_commentsModel = new QStandardItemModel(this);
    m_commentsModel->setHorizontalHeaderLabels({"Offset", "Text"});
    m_commentsModel->setColumnCount(2);
    m_commentsTable->setModel(m_commentsModel);
    m_commentsTable->horizontalHeader()->setStretchLastSection(true);
    m_commentsTable->horizontalHeader()->setSelectionBehavior(QAbstractItemView::SelectRows);
    m_commentsTable->horizontalHeader()->setSelectionMode(QAbstractItemView::SingleSelection);
    m_commentsTable->horizontalHeader()->setEditTriggers(QAbstractItemView::NoEditTriggers);
    m_commentsTable->verticalHeader()->hide();
    m_commentsTable->horizontalHeader()->hide();
    tabWidget->addTab(m_commentsTable, "Comments");

    // Set up variables tab
    // m_variablesTable->setModel(new QStandardItemModel(this));
    // TODO: Add variables to data.
    // tabWidget->addTab(m_variablesTable, "Variables");

    // Add the tab widget to this widget's layout
    QVBoxLayout *layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(0);
    layout->addWidget(tabWidget);

    setLayout(layout);
}

void WarpFunctionInfoWidget::UpdateInfo()
{
    m_commentsModel->clear();
    m_commentsModel->setRowCount(0);
    m_constraintsTable->SetMatchedConstraints({});
    m_constraintsTable->SetConstraints({});

    Warp::Ref<Warp::Function> function = GetFunction();
    if (!function)
        return;

    // Set the analysis constraints if there is an analysis function.
    if (const auto analysisFunction = GetAnalysisFunction())
    {
        const auto analysisConstraints = Warp::Function::Get(*analysisFunction)->GetConstraints();
        m_constraintsTable->SetMatchedConstraints(analysisConstraints);
    }

    // Add all the constraints for the current function to the model.
    QVector<WarpConstraintItem *> constraints;
    for (const auto &constraint: function->GetConstraints())
        constraints.push_back(new WarpConstraintItem(constraint));
    m_constraintsTable->SetConstraints(constraints);

    // Add all the comments to the model.
    for (const auto &comment: function->GetComments())
    {
        m_commentsModel->appendRow({
            new QStandardItem(QString("0x%1").arg(comment.offset, 0, 16)),
            new QStandardItem(QString::fromStdString(comment.text))
        });
    }
}
