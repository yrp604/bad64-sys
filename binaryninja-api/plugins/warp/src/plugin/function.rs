use crate::cache::{cached_function_guid, try_cached_function_guid};
use crate::{get_warp_include_tag_type, INCLUDE_TAG_NAME};
use binaryninja::background_task::BackgroundTask;
use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::command::{Command, FunctionCommand};
use binaryninja::function::Function;
use binaryninja::rc::Guard;
use rayon::iter::ParallelIterator;
use std::thread;
use warp::signature::function::FunctionGUID;

pub struct IncludeFunction;

impl FunctionCommand for IncludeFunction {
    fn action(&self, view: &BinaryView, func: &Function) {
        let sym_name = func.symbol().short_name();
        let sym_name_str = sym_name.to_string_lossy();
        let should_add_tag = func.function_tags(None, Some(INCLUDE_TAG_NAME)).is_empty();
        let insert_tag_type = get_warp_include_tag_type(view);
        match should_add_tag {
            true => {
                log::info!(
                    "Including selected function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.add_tag(&insert_tag_type, "", None, false, None);
            }
            false => {
                log::info!(
                    "Removing included function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.remove_tags_of_type(&insert_tag_type, None, false, None);
            }
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        // TODO: Only allow if the function is named?
        true
    }
}

pub struct CopyFunctionGUID;

impl FunctionCommand for CopyFunctionGUID {
    fn action(&self, _view: &BinaryView, func: &Function) {
        let Some(guid) = cached_function_guid(func, || func.lifted_il().ok()) else {
            log::error!("Could not get guid for copied function");
            return;
        };
        log::info!(
            "Function GUID for {:?}... {}",
            func.symbol().short_name(),
            guid
        );
        if let Ok(mut clipboard) = arboard::Clipboard::new() {
            let _ = clipboard.set_text(guid.to_string());
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        true
    }
}

pub struct FindFunctionFromGUID;

impl Command for FindFunctionFromGUID {
    fn action(&self, view: &BinaryView) {
        let Some(guid_str) = binaryninja::interaction::get_text_line_input(
            "Function GUID",
            "Find Function from GUID",
        ) else {
            return;
        };

        let Ok(searched_guid) = guid_str.parse::<FunctionGUID>() else {
            log::error!("Failed to parse function guid... {}", guid_str);
            return;
        };

        log::info!("Searching functions for GUID... {}", searched_guid);
        let funcs = view.functions();
        let view = view.to_owned();
        thread::spawn(move || {
            let background_task = BackgroundTask::new(
                &format!("Searching functions for GUID... {}", searched_guid),
                false,
            );

            // Only run this for functions which have already generated a GUID.
            let matched: Vec<Guard<Function>> = funcs
                .par_iter()
                .filter(|func| {
                    try_cached_function_guid(func).is_some_and(|guid| guid == searched_guid)
                })
                .collect();

            if matched.is_empty() {
                log::info!("No matches found for GUID... {}", searched_guid);
            } else {
                for func in &matched {
                    // Also navigate the user, as that is probably what they want.
                    if matched.len() == 1 {
                        let current_view = view.file().current_view();
                        if view
                            .file()
                            .navigate_to(&current_view, func.start())
                            .is_err()
                        {
                            log::error!(
                                "Failed to navigate to found function 0x{:0x} in view {}",
                                func.start(),
                                current_view
                            );
                        }
                    }
                    log::info!("Match found at function... 0x{:0x}", func.start());
                }
            }

            background_task.finish();
        });
    }

    fn valid(&self, _view: &BinaryView) -> bool {
        true
    }
}
