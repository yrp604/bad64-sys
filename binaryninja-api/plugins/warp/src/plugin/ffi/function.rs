use crate::build_function;
use crate::cache::{insert_cached_function_match, try_cached_function_match};
use crate::convert::{to_bn_symbol_at_address, to_bn_type};
use crate::plugin::ffi::{BNWARPConstraint, BNWARPFunction, BNWARPFunctionGUID};
use binaryninja::function::Function;
use binaryninja::rc::Ref;
use binaryninja::string::BnString;
use binaryninjacore_sys::{BNFunction, BNSymbol, BNType};
use std::ffi::c_char;
use std::mem::ManuallyDrop;
use std::sync::Arc;
use warp::signature::comment::FunctionComment;

#[repr(C)]
pub struct BNWarpFunctionComment {
    pub text: *mut c_char,
    pub offset: i64,
}

impl BNWarpFunctionComment {
    /// Leaks the text string to be freed with BNWARPFreeFunctionComment
    pub fn from_owned(value: &FunctionComment) -> Self {
        let text = BnString::into_raw(BnString::new(&value.text));
        Self {
            text,
            offset: value.offset,
        }
    }

    pub unsafe fn free_raw(value: &Self) {
        unsafe { BnString::free_raw(value.text) }
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match build_function(&function, || function.lifted_il().ok(), false) {
        Some(function) => Arc::into_raw(Arc::new(function)) as *mut BNWARPFunction,
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetMatchedFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match try_cached_function_match(&function) {
        Some(matched_function) => {
            let arc_matched_function = Arc::new(matched_function);
            // NOTE: Freed by BNWARPFreeFunctionReference
            Arc::into_raw(arc_matched_function) as *mut BNWARPFunction
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionApply(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) {
    let analysis_function = Function::from_raw(analysis_function);
    match function.is_null() {
        false => {
            // Set the matched function to `function`.
            let matched_function = ManuallyDrop::new(Arc::from_raw(function));
            insert_cached_function_match(
                &analysis_function,
                Some(matched_function.as_ref().clone()),
            )
        }
        true => {
            // We are removing the previous match.
            insert_cached_function_match(&analysis_function, None)
        }
    };
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetGUID(
    function: *mut BNWARPFunction,
) -> BNWARPFunctionGUID {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    function.guid
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbol(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNSymbol {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let view = analysis_function.view();
    let address = analysis_function.symbol().address();
    let function_symbol = to_bn_symbol_at_address(&view, &function.symbol, address);
    // NOTE: The symbol ref has been pre-incremented for the caller.
    Ref::into_raw(function_symbol).handle
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbolName(function: *mut BNWARPFunction) -> *mut c_char {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let bn_name = BnString::new(&function.symbol.name);
    // NOTE: The symbol name string to be freed by BNFreeString
    BnString::into_raw(bn_name)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetType(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNType {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    match &function.ty {
        Some(func_ty) => {
            let arch = analysis_function.arch();
            let function_type = to_bn_type(Some(arch), func_ty);
            // NOTE: The type ref has been pre-incremented for the caller.
            unsafe { Ref::into_raw(function_type) }.handle
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetConstraints(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWARPConstraint {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_constraints: Box<[BNWARPConstraint]> = function
        .constraints
        .clone()
        .into_iter()
        .map(Into::into)
        .collect();
    *count = raw_constraints.len();
    let raw_constraints_ptr = Box::into_raw(raw_constraints);
    raw_constraints_ptr as *mut BNWARPConstraint
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetComments(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWarpFunctionComment {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_comments: Box<[_]> = function
        .comments
        .iter()
        .map(BNWarpFunctionComment::from_owned)
        .collect();
    *count = raw_comments.len();
    let raw_comments_ptr = Box::into_raw(raw_comments);
    raw_comments_ptr as *mut BNWarpFunctionComment
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionsEqual(
    function_a: *mut BNWARPFunction,
    function_b: *mut BNWARPFunction,
) -> bool {
    // We do not own function so we should not drop.
    let function_a = ManuallyDrop::new(Arc::from_raw(function_a));
    // We do not own function so we should not drop.
    let function_b = ManuallyDrop::new(Arc::from_raw(function_b));
    function_a.eq(&function_b)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPNewFunctionReference(
    function: *mut BNWARPFunction,
) -> *mut BNWARPFunction {
    Arc::increment_strong_count(function);
    function
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionReference(function: *mut BNWARPFunction) {
    if function.is_null() {
        return;
    }
    Arc::decrement_strong_count(function);
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionList(functions: *mut *mut BNWARPFunction, count: usize) {
    let functions_ptr = std::ptr::slice_from_raw_parts_mut(functions, count);
    let functions = Box::from_raw(functions_ptr);
    for function in functions {
        // NOTE: The functions themselves should also be arc.
        BNWARPFreeFunctionReference(function);
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionCommentList(
    comments: *mut BNWarpFunctionComment,
    count: usize,
) {
    let comments_ptr = std::ptr::slice_from_raw_parts_mut(comments, count);
    let comments = Box::from_raw(comments_ptr);
    for comment in &comments {
        BNWarpFunctionComment::free_raw(comment)
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeConstraintList(
    constraints: *mut BNWARPConstraint,
    count: usize,
) {
    let constraints_ptr = std::ptr::slice_from_raw_parts_mut(constraints, count);
    let _constraints = unsafe { Box::from_raw(constraints_ptr) };
}
