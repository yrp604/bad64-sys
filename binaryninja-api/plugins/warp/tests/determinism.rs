//! This tests to make sure the function GUIDs are stable.
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use std::collections::BTreeMap;
use std::path::PathBuf;
use warp::signature::function::FunctionGUID;
use warp_ninja::cache::cached_function_guid;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 10] = [
    "_ctype.obj",
    "_fptostr.obj",
    "_mbslen.obj",
    "_memicmp.obj",
    "_strnicm.obj",
    "_wctype.obj",
    "atof.obj",
    "atoldbl.obj",
    "atox.obj",
    "ls",
];

#[test]
fn insta_signatures() {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");
        let functions: BTreeMap<u64, FunctionGUID> = view
            .functions()
            .iter()
            .map(|f| {
                let guid = cached_function_guid(&f, || f.lifted_il().ok());
                (f.start(), guid.unwrap())
            })
            .collect();
        insta::assert_debug_snapshot!(file_name, functions);
    }
}
