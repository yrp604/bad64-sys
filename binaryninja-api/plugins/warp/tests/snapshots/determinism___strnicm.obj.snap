---
source: plugins/warp/tests/determinism.rs
expression: functions
---
{
    65728: FunctionGUID {
        guid: 2e79ac68-8b77-500f-9c36-871e9f7bc63a,
    },
}
